pkgs/_test_macros/lib/literal_params.dart

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:dart_model/dart_model.dart';
+// ignore: implementation_imports
+import 'package:dart_model/src/macro_metadata.g.dart';
+import 'package:macro/macro.dart';
+import 'package:macro_service/macro_service.dart';
+
+import 'templating.dart';
+
+/// Covers macro metadata cases where the params will always be written as
+/// literals in the annotation.
+///
+/// Outputs comments with evaluation results.
+///
+/// Throws if the annotation has something other than supported literals.
+/// TODO(davidmorgan): support diagnostics, make failures diagnostics.
+class LiteralParams {
+  final int? anInt;
+  final num? aNum;
+  final double? aDouble;
+  final String? aString;
+  final Object? anObject;
+  final List<int>? ints;
+  final List<num>? nums;
+  final List<double>? doubles;
+  final List<String>? strings;
+  final List<Object>? objects;
+
+  const LiteralParams(
+      {required this.anInt,
+      this.aNum,
+      this.aDouble,
+      this.aString,
+      this.anObject,
+      this.ints,
+      this.nums,
+      this.doubles,
+      this.strings,
+      this.objects});
+}
+
+class LiteralParamsImplementation implements ClassDeclarationsMacro {
+  // TODO(davidmorgan): this should be injected by the bootstrap script.
+  @override
+  MacroDescription get description => MacroDescription(
+      annotation: QualifiedName(
+          uri: 'package:_test_macros/literal_params.dart',
+          name: 'LiteralParams'),
+      runsInPhases: [2]);
+
+  @override
+  Future<void> buildDeclarationsForClass(
+      ClassDeclarationsBuilder builder) async {
+    final annotation = builder
+        .target.metadataAnnotations.first.expression.asConstructorInvocation;
```

Comment:
We should probably add a TODO for checking the type of the annotation (in practice you will want to search for a specific one).

Reply:
Good point, thanks! Done.